docs/migrating-from-0.2-to-1.0.md

```diff
@@ -372,3 +374,59 @@ The `embedded-hal` project now spans several crates, where some functionality ha
 | [embedded-io-async](./embedded-io-async)       | [![crates.io](https://img.shields.io/crates/v/embedded-io-async.svg)](https://crates.io/crates/embedded-io-async) | [![Documentation](https://docs.rs/embedded-io-async/badge.svg)](https://docs.rs/embedded-io-async) | I/O traits, async version  |
 | [embedded-io-adapters](./embedded-io-adapters)       | [![crates.io](https://img.shields.io/crates/v/embedded-io-adapters.svg)](https://crates.io/crates/embedded-io-adapters) | [![Documentation](https://docs.rs/embedded-io-adapters/badge.svg)](https://docs.rs/embedded-io-adapters) | Adapters between the [`embedded-io`](https://crates.io/crates/embedded-io) and [`embedded-io-async`](https://crates.io/crates/embedded-io-async) traits and other IO traits (`std`, `tokio`, `futures`...)  |
 
+## Supporting both 0.2 and 1.0 in the same HAL
+
+It is strongly recommended that HAL implementation crates provide implementations for both the `embedded-hal` v0.2 and v1.0 traits.
+This allows users to use drivers using either version seamlessly.
+
+The way you do it is adding a dependency on both versions in `Cargo.toml` like this:
+
+```toml
+[dependencies]
+embedded-hal-02 = { package = "embedded-hal", version = "0.2.7", features = ["unproven"] }
+embedded-hal-1 = { package = "embedded-hal", version = "1.0" }
+```
+
+This allows you to refer to the v0.2 traits under the `embedded_hal_02` name, and the v1.0 traits under
+`embedded_hal_1`. Implement both versions on the same struct. For example, for an input pin:
+
+```rust
+/// The HAL's input pin struct
+struct Input {...}
+
+/// Implement the v0.2 traits on the struct.
+impl embedded_hal_02::digital::v2::InputPin for Input {
+    type Error = Infallible;
+
+    fn is_high(&self) -> Result<bool, Self::Error> {
+        ...
+    }
+
+    fn is_low(&self) -> Result<bool, Self::Error> {
+        ...
+    }
+}
+
+/// ... and implement the v1.0 traits on the *same* struct.
+impl embedded_hal_1::digital::ErrorType for Input {
+    type Error = Infallible;
+}
+
+impl embedded_hal_1::digital::InputPin for Input {
+    fn is_high(&mut self) -> Result<bool, Self::Error> {
+        ...
+    }
+
+    fn is_low(&mut self) -> Result<bool, Self::Error> {
+        ...
+    }
+}
```

Comment:
May be better propose v1.0 as main implementation and v.0.2 as dependent.
```Rust
      /// Implement the v1.0 traits on the struct.
    impl embedded_hal_1::digital::ErrorType for Input {
        type Error = Infallible;
    }

    impl embedded_hal_1::digital::InputPin for Input {
        fn is_high(&mut self) -> Result<bool, Self::Error> {
            ...
        }

        fn is_low(&mut self) -> Result<bool, Self::Error> {
            ...
        }
    }

    /// Implement the v0.2 traits on the struct.
    impl<P: embedded_hal_1::digital::InputPin> embedded_hal_02::digital::v2::InputPin for P {
        type Error = <Self as embedded_hal_1::digital::ErrorType>::Error;

        fn is_high(&self) -> Result<bool, Self::Error> {
            self.is_high()
        }

        fn is_low(&self) -> Result<bool, Self::Error> {
            self.is_low()
        }
    }
```
v0.2 also might be under the some feature

Reply:
> May be better propose v1.0 as main implementation and v.0.2 as dependent.

i'd rather leave that up to the HAL, there's a few ways of doing it. For example, some HALs have inherent methods so the user can call them without importing any traits (and so the API can be different, for example for GPIO it's nice to make it infallible). In that case the trait methods would call into the inherent methods.

(In this example I just wanted to show how to add the two versions as dependencies, as it's relatively unknown that it's possible. I've seen people entirely remove 0.2 just because they assumed it's not)

> v0.2 also might be under the some feature

Same, it's better left up to each HAL.